File: articles/linear-algebra-for-systems-modeling/articles/latent-structure-and-signal-extraction/calculators/r/latent_structure_calculator.R
result <- data.frame(
  calculator = "latent_structure_and_signal_extraction_calculator",
  model_name = "synthetic_latent_structure_signal_extraction_audit",
  observations = 9,
  variables = 6,
  method = "svd_low_rank_signal_extraction",
  preprocessing = "centered_and_standardized",
  retained_rank = 2,
  retained_signal_ratio = 0.962,
  relative_reconstruction_error = 0.195,
  maximum_observation_residual = 1.43,
  highest_residual_observation = 8,
  warning = "Latent signal extraction metrics depend on observed matrix construction, preprocessing, method choice, retained rank, signal definition, residual review, stability validation, and interpretation context."
)

dir.create("outputs", recursive = TRUE, showWarnings = FALSE)
write.csv(result, "outputs/r_latent_structure_and_signal_extraction_calculator.csv", row.names = FALSE)
print(result)

File: articles/linear-algebra-for-systems-modeling/articles/latent-structure-and-signal-extraction/calculators/python/latent_structure_calculator.py
from __future__ import annotations

import csv
import json
from pathlib import Path


def main() -> None:
    output_dir = Path("outputs")
    output_dir.mkdir(parents=True, exist_ok=True)

    result = {
        "calculator": "latent_structure_and_signal_extraction_calculator",
        "model_name": "synthetic_latent_structure_signal_extraction_audit",
        "observations": 9,
        "variables": 6,
        "method": "svd_low_rank_signal_extraction",
        "preprocessing": "centered_and_standardized",
        "retained_rank": 2,
        "retained_signal_ratio": 0.962,
        "relative_reconstruction_error": 0.195,
        "maximum_observation_residual": 1.43,
        "highest_residual_observation": 8,
        "warning": "Latent signal extraction metrics depend on observed matrix construction, preprocessing, method choice, retained rank, signal definition, residual review, stability validation, and interpretation context."
    }

    with (output_dir / "latent_structure_and_signal_extraction_calculator.json").open("w", encoding="utf-8") as handle:
        json.dump(result, handle, indent=2, sort_keys=True)

    with (output_dir / "latent_structure_and_signal_extraction_calculator.csv").open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(result.keys()))
        writer.writeheader()
        writer.writerow(result)

    print(json.dumps(result, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
